Audio playback refactor

seadowg opened the pull request:
Closes #2521

There's already been some good work and discussion around how to fix some of the bugs and quirks around audio playback in the app (#2522, #2863 and #2817 for example). For people not familiar with the bugs: there are multiple problems with play/stop buttons interacting with one another. After having a look at the problem, it made sense to give it another stab now that we have some new tooling maturing in the Android Jetpack world.

After digging into the current architecture it made sense to me that a good way forward would be to separate out audio playback functionality entirely from the question world so that it could be tested and implemented in isolation. To help myself out I ended up drawing a picture of this in [Miro](https://miro.com):

<img width="1037" alt="Screenshot 2019-07-23 at 10 20 28" src="https://user-images.githubusercontent.com/556280/61699838-881c3c80-ad33-11e9-918f-4e19827ae24b.png">

This draft starts with an initial pass at that that uses an `AudioPlayerViewModel` to coordinate the `MediaPlayer` and expose state and actions to an `AudioButton` view that is only responsible for rendering whether it is in the "playing" state or not and for emitting (via a listener) "playClicked" and "stopClicked" events.

I worked through this with @lognaturel and one thing we noticed was how many different behaviors of the app that audio playback is part of. Although this isn't finished yet it made sense to get a draft PR up early to get feedback and also have people point out features we might be missing. I'm going to enumerate them below (and tick off what is working in this reimplementation) - shout out if anything is missing:

- [x] Question audio can be played/stopped
- [x] Select option audio can be played/stopped
- [x] Playing one audio button on a screen stops any other playing one
- [x] Question option text is highlight when audio is being played
- [x] Select option text is highlighted when audio is being played
- [x] `MediaPlayer` objects are released when `Activity` is destroyed
- [x] `MediaPlayer` objects are released when `Activity` is paused
- [x] Moving between form screens stops playback
- [x] Audio can be autoplayed for questions that are being **forward** swiped into
- [x] Playing an audio button should stop an `AudioWidget` (and vice versa)
- [x] Audio in an audio widget can be played/paused
- [x] Audio in an audio widget can be seeked with buttons
- [x] Audio in an audio widget can be seeked with seek bar
- [x] Pausing/resuming/rotating Activity doesn't cause any problems (currently playing audio etc)

**Just to note**: this PR does not fix the problems outlined in #2861 - autoplay for select options is not functioning correctly. We'll follow up with a fix for that once this is in.

#### What has been done to verify that this works as intended?

Used several different forms involving `AudioWidget`s, audio in question media and in select options etc etc.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

There might be some accidental changes to behavior given how little of this functionality was tested or documented. It would be to double check that media in forms generally still works as expected (I've listed the behaviours I could find above) and that we don't see any (negative) performance changes compared to the current version of the app. This is a big set of changes so I wouldn't be surprised if find problems - I'd imagine with something I didn't realize the app could do.

#### Do we need any specific form for testing your changes? If so, please attach one.

Good to use as many media forms as possible but a good one to start with is the "MediaForm" linked in #2521.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

I think it would be good to add documentation for the audio functionality. I'll make sure to file an an issue once we're happy with this!

#### Before submitting this PR, please make sure you have:
- [ ] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

seadowg committed:
Spike out AudioPlayerViewModel

seadowg committed:
Move audio components to their own package

seadowg committed:
Use reworked audio button in MediaLayout

seadowg committed:
Remove coupling from AudioButton to ViewModel

seadowg committed:
Make PlayStop listener methods a little more obvious

seadowg committed:
Use seperate ID for audio playing state

This prevents problems where two buttons use the same audio file.

seadowg committed:
Clean up AudioButton class

seadowg committed:
Make ReferenceManager an explicit dependency of MediaLayout

seadowg committed:
Add field for reference manager to QuestionWidget

seadowg committed:
Inject ReferenceManager into QuestionWidget

I don't think this is the ideal solution as bringin Dagger into
a View is potentially overkill but the alternative is to add a
constructor argument which I think is a change best made on it's
own as so many widgets will be affected.

seadowg committed:
Highlight question text when audio is playing

seadowg committed:
Small refactor to AudioButtonIntegrationTest

seadowg committed:
Remove references to unused divider

seadowg committed:
Move text highlighting test to MediaLayout from QuestionWidget

This is the actual component shared between all uses of this feature

seadowg committed:
Remove MediaPlayer from MediaLayout

seadowg committed:
Move away from static method

You're welcome future person who as to write a test around this boundary.

seadowg committed:
Move AudioController to audio package

seadowg committed:
Rename AudioController layout

seadowg committed:
Rename AudioController layout

seadowg committed:
Release media player on Activity pause

seadowg committed:
Add pause to AudioPlayerViewModel

seadowg commented:
Interesting note from something I've realised: the way that swiping through pages/groups works in Collect means that it's hard for us to use the `FormEntryActivity` as our "lifecycle" for architecture components. The lifecycle of most views is really dictated by their parent `ODKView` as they are created and destroyed as we move through each group of questions. This suggests that we should really be using a `Fragement` to own each group (which would also let us use components like the `ViewPager`. That's certainly a change for another time though so for now we may just have to give the `ODKView` its own `Lifecycle` object - this will be a little hacky but doable I think.

**EDIT**: A way nicer way to handle this is to just have the `FormEntryActivity` notify the `AudioPlayerViewModel` to stop the `MediaPlayer` we can use the fact that our `ViewModel` is cached at the Activity level to make this happen easily.

seadowg committed:
Use AndroidScreen interface to prevent coupling to FormEntryActivity

seadowg committed:
Revert "Use AndroidScreen interface to prevent coupling to FormEntryActivity"

This reverts commit 62b45c6a2feb850ecc6ddc772d3ab989964e713d.

seadowg commented:
Our "nicer way" in my last comment only solves part of the problem as we still need a valid lifecycle for `LiveData` observations. I'm going to spike/investigate 3 paths forward and report back with interestings when I'm done:


- [x] Give the `ODKView` it's own "lifecycle" using the lifecycle tools Jetpack gives us.
- [x] Remove lifecycle from observations and manage deregistering ourselves
- [x] Make the `ODKView` a `Fragment` so it has lifecycle we can use

seadowg commented:
Ok here's my findings from the [spikes](https://en.wikipedia.org/wiki/Spike_%28software_development%29):

**1. Give the ODKView it's own "lifecycle" using the Lifecycle tools Jetpack gives us.**

In the `FormEntryActivity` I ended up using an `ObserverToken` (an object I made that implements `LifecycleOwner`). Here's the definition of `ObserverToken` and how it's used in the `FormEntryActivity`:

```java
// ObserverToken

public class ObserverToken implements LifecycleOwner {

    private LifecycleRegistry lifecycleRegistry = new LifecycleRegistry(this);

    public ObserverToken() {
        lifecycleRegistry.handleLifecycleEvent(ON_RESUME);
    }

    public void release() {
        lifecycleRegistry.handleLifecycleEvent(ON_DESTROY);
    }

    @NonNull
    @Override
    public Lifecycle getLifecycle() {
        return lifecycleRegistry;
    }
}


// FormEntryActivity

private ObserverToken odkViewObserverToken = new ObserverToken();

public ObserverToken getODKViewObserverToken() {
    return odkViewObserverToken;
}

private void releaseOdkView() {
    odkViewObserverToken.release();
    odkViewObserverToken = new ObserverToken();
    
    if (odkView != null) {
        odkView.releaseWidgetResources();
        odkView = null;
    }
}
```

You can see here that we end up with the `ObserverToken` standing in as "lifecycle" for the `ODKView`. Then we can use it in views like the `MediaLayout`:

```java
LifecycleOwner lifecycleOwner = activity.getODKViewObserverToken();
liveData.observe(lifeCycleOwner, () -> ...);
```

**2. Remove lifecycle from observations and manage deregistering ourselves.**

Here I ended up with a `LiveDataObserver` object that we can instantiate for classes that observer  `LiveData` but have an "out of sync" lifecycle (like the views under `ODKView`). Here's the definition of `LiveDataObserver` and how it's used:

```java
// LiveDataObserver

public class LiveDataObserver {

    private final List<Pair<LiveData, Observer>> observations = new ArrayList<>();

    public <T> void observe(LiveData<T> liveData, Observer<T> observer) {
        liveData.observeForever(observer);
        observations.add(new Pair<>(liveData, observer));
    }

    public void release() {
        for (Pair<LiveData, Observer> observation : observations) {
            LiveData liveData = observation.first;
            Observer observer = observation.second;

            liveData.removeObserver(observer);
        }

        observations.clear();
    }
}

/// MediaLayout (and anywhere we observer LiveData)

private LiveDataObserver liveDataObserver = new LiveDataObserver();

liveDataObserver.observe(liveData, () -> ...);

public void release() {
    liveDataObserver.release();
}
```

Here you can see that we need to be able to call a method on any view that uses LiveData
to release the observers. This is doable though as we can start at `ODKView.releaseWidgetResources()` and work our way down the view tree from there.

**3. Make the ODKView a Fragment so it has lifecycle we can use.** 

It may not surprising to hear this but: this seems like it's not going to be easy. Most likely we'd need to go for a full ViewPager/Fragment switch over. This would involve moving widget rendering code into a Fragment. I spent a bit of time on playing around with this but pulled out when it looked like it was going to end up being the rest of my week.

### Recommendation

Option 2 leaves us with the ability to use `LiveData` later but removes a lot of the benefits of it (the lifecycle aware parts) while Option 1 lets us use `LiveData` properly while introducing some weirdness to the `FormEntryActivity`. To me it feels like Option 2 is essentially us adapting new code for old problems while Option 1 is the reverse of that. In my opinion, it makes sense to go with Option 1 as long as we're looking to move the `FormEntryActivity` to a ViewPager setup (it seems like there may be some UX advantages to this as well as just simpler code). This would mean that our slightly weird `ObserverToken` solution would just be temporary and would mean that the new Audio feature code wouldn't have to change when the Fragment is introduced.

We could of course drop `LiveData` entirely but, to me, it does feel like this part of the app heavily benefits from a reactive framework of some kind. Rx is of course the standard alternative but then we're back in the world of lifecycle management anyway.

It'd be great to get your thoughts on this @lognaturel and @grzesiek2010!

seadowg committed:
Make sure MediaPlayer is released between ODKView swipes

seadowg committed:
Move clipID generation out of audio package

seadowg committed:
Use mock AudioButtonManager in MediaLayoutTest

lognaturel commented:
Thanks for the clear presentation of your spike results, @seadowg. I'm convinced that if we're going to go in on `LiveData`, the general approach of representing the creation and destruction of the questions views as lifecycles seems preferable to using `observeForever`. Also agree that also doing the refactor to Fragments here will make this too big and that it would be ideal to do it in stages.

Forgive me if this should be obvious, but I'm not clear on why you need to add the `ObserverToken` concept. Couldn't `ODKView` implement `LifecycleOwner`, handle `ON_RESUME` in its constructor and handle `ON_DESTROY` in `releaseWidgetResources` (which should then be renamed)? What am I missing?

seadowg commented:
@lognaturel Yeah that would totally work. My thinking when I was spiking things out was just keeping things as contained as possible so it would be easier to present them back - that's how I ended up with the token rather than it in the view code. Moving forward though it might be a nice way to actually implement it as it keeps all the Jetpack lifecycle code out of the view itself which makes it feel a little more "bolt on" (and hopefully "bolt off"). What do you think?

seadowg committed:
Pass ScreenContext to AudioButtonManager rather than Activity

This lets us seperate the lifecycle from the Activity itself without
creating a dependency on the FormEntryActivity

seadowg committed:
Create a view lifecycle for ODKView

seadowg committed:
Move instance dependencies to AudioButtonManager constructor

seadowg committed:
Rename AudioButtonManager to AudioHelper

seadowg committed:
Move Audio components back to being package private

lognaturel commented:
> keeping things as contained as possible

Ah yes, that did make for a really nice summary. 👍 I'm generally very much for compartmentalization but in this case, I wonder whether having the `ODKView` have the lifecycle field would make things easier to understand and closer to the end goal of having a Fragment be the container instead. 

I don't feel super strongly about it but I am liking what I see in 6f1044a. It feels a little funky to have something that's a Lifecycle**Owner** be called `odkViewLifecycle` but I think it makes enough sense.

seadowg commented:
@lognaturel yeah I think I agree. On second thoughts I think this might also fit in to the `FormEntryActivity` in a neater way as well as we won't have a second field to manage. I'll have a go at making the `ODKView` implement `LifecycleOwner` and see how it looks!

seadowg committed:
Use formIndex as clipID for audio

seadowg commented:
@lognaturel ach I'd forgotten (weekend brain 🤷‍♂️) that it's awkward to have the lifecycle on the `ODKView` itself as the widgets are created as part of the `ODKView` construction. This means the `ODKView` itself hasn't been assigned to the `FormEntryActivity` field. We could rework this all but I suggest we leave it for the moment as that would be better incorporated as a step on the road to "Fragmentizing" the `ODKView`.

seadowg committed:
Initial spiked implementation of autoplay for questions

seadowg committed:
Add helper to encapsulate autoplay logic

seadowg committed:
Add test for AudioButton

seadowg committed:
Make AudioControllerView a View

seadowg committed:
Remove AudioControllerView dependency on FormEntryActivity

seadowg committed:
Suppress find bugs for Butterknifed field

seadowg committed:
Use AudioHelper to setup AudioControllerView

seadowg committed:
Add total and current duration to AudioControllerView

seadowg committed:
Move zip to transformations helper class

seadowg committed:
Use ScreenContextFragement in WidgetTest classes

This is an improvement on using FormEntryActivity as we don't need
to worry about the lifecycle problems and also decouple the widget tests
from the FormEntryActivity. Additionally, it looks like
we get a speed up on tests (I save about 1 minute locally) from not
running the `onCreate` in FormEntryActivity.

seadowg committed:
Cancel timed duration updates when ViewModel is cleared

seadowg committed:
Update seekbar as audio plays on AudioControllerView

lognaturel commented:
@seadowg and I have talked through this a couple of times and I'm on board with the current direction. The big thing I'd like to see in a cleanup pass is more narration (is anyone surprised). For example, capturing some of the back and forth we had above about where the lifecycle used for the `ODKView` should go would be valuable. I think `ViewHelper` and `ScreenContext` are particular candidates for some high-level description of why things are as they are now and where they might go in an ideal future.

I'm going to be out for the next two weeks. @grzesiek2010, please take over review! Don't hesitate to get on a call to discuss. I think it'd be good for this to go into QA as soon as possible because the big risk is that there's functionality missing. @seadowg says he'll likely have something ready for that mid-week. 🙏

seadowg committed:
Add fast fwd and rewind buttons

seadowg committed:
Make sure position on audio view model is within file bounds

seadowg committed:
Remove FormEntryActivity from more widget tests

seadowg committed:
Swiping bar in AudioWidget seeks playback

seadowg committed:
Make swiping behave correctly for multiple audio widgets

seadowg committed:
Simplify audio implementation by moving position state into view model

seadowg committed:
Make sure seek bar doesn't disrupt swiping in forms

seadowg committed:
Remove unneeded variable

seadowg committed:
Correct theme name

seadowg committed:
Add docs for Scheduler

seadowg committed:
Add docs to ScreenContext

seadowg committed:
Sever dependency between AudioHelper and ScreenContext

seadowg committed:
Add docs to AudioHelper

seadowg committed:
Simplify AudioControllerView by removing NOT_PLAYING state

seadowg committed:
Replace TimerScheduler with HandlerScheduler

This is based on a lot of literature around Handler being a better choice
https://stackoverflow.com/questions/20330355/timertask-or-handler

seadowg committed:
Remove unused code

grzesiek2010 commented:
Hi I'm in the middle of reviewing/testing and here is what I have already found:

1. Weird error when I open a form which uses autoplay option but there is no audio file specified:
![Screenshot_2019-08-19-14-01-31](https://user-images.githubusercontent.com/3276264/63263902-111e8900-c28a-11e9-93c4-f76a274d3143.png)

[autoPlayTest.zip](https://github.com/opendatakit/collect/files/3515763/autoPlayTest.zip)

2. Autoplay doesn't work in `field-list`. Maybe you decided to keep such limitation because it would be too complicated so correct me if it's not a bug.

3. Autoplay doesn't work for audio files attached to choices. Same as above correct me if it's not a bug.

seadowg commented:
@grzesiek2010 thanks for playing around with it!

> Weird error when I open a form which uses autoplay option but there is no audio file specified

Ah nice catch. Do you know what this situation would have resulted in before? I can have a look if not so we can make a call if this is better/worse.

> Autoplay doesn't work in field-list. Maybe you decided to keep such limitation because it would be too complicated so correct me if it's not a bug.

By this do you mean where I have a group of questions on one page - like in your "MediaForm" example? I'm guessing you're thinking it could play each questions audio in order. This functioanlity wasn't working before was it? This wasn't something that I considered. Maybe it will be something we see a need for in the discussions in #2861.

> Autoplay doesn't work for audio files attached to choices. Same as above correct me if it's not a bug.

This was considered but for this my plan was to follow up with that later as it wasn't currently working. I added that the description of the PR but it's a little hidden - I've bolded it to make it more obvious!

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/audio/AudioPlayerViewModel.java:
```diff
@@ -0,0 +1,190 @@
+        }
+    }
+
+    private enum ClipState {
```
Seems unused.

seadowg replied:
Good catch it is. Thought PMD/CheckStyle would have caught that! Will delete.

seadowg committed:
Remove unused code

seadowg commented:
@grzesiek2010 it looks like the current release of the app crashes when there is a missing audio URI while trying to autoplay. The new behaviour of showing a weird message in a dialog and then proceeding to the next screen of the form isn't great either though.

To me it makes sense for this to just fail silently and display the question/field list. What do you think?

seadowg committed:
Silently fail when audio URI missing for autoplay

grzesiek2010 commented:
Yes, it should just be ignored if a corresponding media file is missing.

seadowg commented:
@grzesiek2010 cool. I've added a commit with that change.

grzesiek2010 commented:
> By this do you mean where I have a group of questions on one page - like in your "MediaForm" example? I'm guessing you're thinking it could play each questions audio in order. This functioanlity wasn't working before was it?

yes but I'm not sure if it worked before it could a long time ago. I think it's not important and we can leave it as-is for now.

> This was considered but for this my plan was to follow up with that later as it wasn't currently working. I added that the description of the PR but it's a little hidden - I've bolded it to make it more obvious!

ok, I think it's not important for now. Maybe we don't need it at all.

grzesiek2010 commented:
ok, so I'm not going to test more let's ask @mmarciniak90 to do that.
@mmarciniak90 the functionality has never worked as expected and might be full of some tricky cases like the one I caught so be careful and take all the time you need.

seadowg commented:
Found a bug while working on the autoplay features:

1. Open a form with audio media on a question
2. Play the audio and let it finish
3. Press play again

The button makes it look like the audio only plays for a split second and only the end of the audio is heard. I'll remove the `needs-testing` and fix this.

seadowg committed:
Make sure audio button playback resets

seadowg commented:
@mmarciniak90 sorry I forgot to readd the "needs-testing" label to this. Will do.

mmarciniak90 commented:
@seadowg, the first thing that I noticed is that audio is not always stopped when I swipe to the next page. It works for `Birds` form but not for `Media Form`.

seadowg commented:
@mmarciniak90 ah I think I see what might be wrong. Is the proble happening on the last screen of the form every time?

seadowg committed:
Make sure ODKView lifecycle is always destroyed when form entry screen changes

seadowg commented:
@mmarciniak90 I've pushed a fix that should hopefully resolve that problem

mmarciniak90 commented:
@seadowg sorry that I'm adding issues partially. I'm still working on testing this PR.

* Crash occurs when **audio** file is missing and the user clicks on `Play` button. Problem is not visible when a video file is missed. Toast with message `File ... is missing` should be visible.

`org.odk.collect.android E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.odk.collect.android, PID: 23123
    java.lang.RuntimeException
        at org.odk.collect.android.audio.AudioPlayerViewModel.loadClip(AudioPlayerViewModel.java:176)
        at org.odk.collect.android.audio.AudioPlayerViewModel.play(AudioPlayerViewModel.java:37)
        at org.odk.collect.android.audio.AudioHelper$AudioButtonListener.onPlayClicked(AudioHelper.java:155)
        at org.odk.collect.android.audio.AudioButton.onClick(AudioButton.java:82)
        at android.view.View.performClick(View.java:6213)
        at android.view.View$PerformClick.run(View.java:23645)
        at android.os.Handler.handleCallback(Handler.java:751)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6692)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1468)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1358)`

* `test_audio` form: autoplay works when the form is loaded and when the user jumps to a question. It does not work when the user **swipes back** to question with autoplay audio

I  also noticed behavior which could be improved but they exist on v1.23.3 and they are related only to Android 8.1 and 9.0
* rotating the device does not stop playing an audio file
* `test_audio` form: video file is played, again and again, it does not return to form

seadowg commented:
@mmarciniak90 no that's great. Good to get these things early! I'll have a look at the crash.

```
test_audio form: autoplay works when the form is loaded and when the user jumps to a question. It does not work when the user swipes back to question with autoplay audio
```

That was the old behaviour (I thought anyway) so I left it intact. The behaviour makes sense to me but maybe we play around with it in future to see if it's what people want?

```
rotating the device does not stop playing an audio file
```

Ah yeah that will work now. Do we think it should stop on rotate?

```
test_audio form: video file is played, again and again, it does not return to form
```

I saw this as well. It seemed on my devices (and emulator) Android was set to "loop video" (it's in the options menu of the video) by default.

mmarciniak90 commented:
@seadowg 
1.  > The behaviour makes sense to me but maybe we play around with it in future to see if it's what people want?

Sounds good

2. I see some inconsistency here because on Android 8.1 and 9 audio is played even if device orientation is changed but on all lower Android versions, audio is stopped.

3. You have right! There was `Loop video` option selected. But still, it does not return to form, to the question. Lower Android versions are moving the user to form when the video ends. 

As I said, the last two points are visible in v1.23.3.

seadowg commented:
> I see some inconsistency here because on Android 8.1 and 9 audio is played even if device orientation is changed but on all lower Android versions, audio is stopped.

> You have right! There was Loop video option selected. But still, it does not return to form, to the question. Lower Android versions are moving the user to form when the video ends.

Both of these feel like they're ok but I'll take a look into **why** they happen to make sure neither are suggesting some underlying problem.

seadowg committed:
Stop app crashing when audio file is missing

seadowg committed:
Show toast when audio error occurs

seadowg commented:
@lognaturel would be interesting to get your opinion of the style of error handling (using `LiveData`) in 418e43f8bf05d2c8e99ef1dba02dd606d8b0db2f

seadowg committed:
Fix compilation errors

lognaturel commented:
These kinds of errors could happen multiple times while a specific question/field list is being displayed so modeling them as events that are streamed seems fine to me. Any alternative I can think of would require the view model or audio helper to have some knowledge of where errors should be displayed. I can't think of an alternative using `LiveData`. What kind of objections were you thinking I might have? Did you consider alternatives?

Where I haven't been able to find an analogous structure I liked is when the info that needs to be displayed is only displayed exactly once. For example, that's the case with the background location tracking warning toast and that's why I ultimately abandoned using a `LiveData`-based solution there.

seadowg commented:
@lognaturel the only alternative I considered was having `play` throw an `Exception`. The disadvantage was having to show toasts from multiple places - wherever `play` is called. I like that this solution let us set up our error handling in one place and then let everything else forget about it.

> when the info that needs to be displayed is only displayed exactly once

Ah that's interesting. In my head you'd do this by only "emitting" the error once from a `ViewModel` as it feels like the fact that the error has happened before becomes applicaiton state rather than view state.

seadowg commented:
@mmarciniak90 I spent some time look into the orientation changes causing the audio to stop. I found the same thing that on API 26+ (8.0) the audio doesn't stop when the phone is rotated but on older devices it does. 

**TLDR**: this is because on devices before API 26 the app does the same thing when it rotates as when we change screens or switch apps and we stop audio at those points.

**Full(er) explanation**: This seems to be because of a behaviour change in Android but is also to do with some potential weirdness on our side. It looks like we're using a flag in our `AndroidManifest.xml` on a lot of our Activities that stops them from being recreated when we go through an orientation change (`configChanges=orientation`). On devices before API 26 however it looks like the `FormEntryActivity` still goes through `onPause`, `onDestroy` etc. Because our new audio implementation stops playback and releases the `MediaPlayer` when it's host `Activity` is paused the playback stops for these API levels. API 26+ devices behave as you'd expect: the `Activity` doesn't pause on rotate and the playback doesn't stop.  If I remove the `configChanges` flag from the `FormEntryActivity` then the behaviour is consistent across API levels (the playback always stops).

Frankly this was really confusing and it took me a while to realize those flags were on the manifest. I couldn't find anything on the behaviour of `configChanges` being altered at API 26 but it does seem that [Google aren't encouraging](https://developer.android.com/guide/topics/resources/runtime-changes#HandlingTheChange) using the flag. It looks like @shobhitagarwal1612 posted this as an issue back in 2017 with #799. I think we should leave this as is for this PR but I would be tempted to investigate removing the `configChanges` flag in future to avoid these kind of inconsistencies.

mmarciniak90 commented:
@seadowg Thanks for the so precise analyze. This knowledge will be helpful for testing other PRs too.

**Tested with success**

**Verified on Android:** 4.2, 4.4, 5.1, 6.0, 7.0, 8.1, 9.0

**Verified cases:**
* light/dark mode
* delete media - `File ... is missing` is visible.
* a few questions on one page - play/pause media from different questions
* highlighting question or answer when media is played
* minimize - media is stopped
* swipe to last view - media is stopped
* rotating when media is playing; Android 8.1 and 9.0 - media is continued; lower versions - audio is stopped

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

seadowg commented:
@lognaturel @grzesiek2010 feel happy with this merging in?

lognaturel commented:
Thrilled! 🎉
